client/store/actionCreators.js

```diff
@@ -10,202 +11,224 @@ import {
+});
+
+const loginFail = (message) => ({
+  type: LOGIN_FAIL,
+  message,
+});
+
+export const loading = () => ({
+  type: LOADING,
+});
+
+export const loaded = () => ({
+  type: LOADED,
+});
+
+const getProducts = (products) => ({
+  type: GET_PRODUCTS,
+  products,
+});
+
+const getCategories = (categories) => ({
+  type: GET_CATEGORIES,
+  categories,
+});
+
+const addToCart = (cart) => ({
+  type: ADD_TO_CART,
+  cart,
+});
+
+const getCart = (cart) => ({
+  type: GET_CART,
+  cart,
+});
+
+export const loginThunk = (username, password) => (dispatch) => {
+  dispatch(loading());
+  return axios
+    .post('/user/login', { username, password })
+    .then((res) => {
+      console.log(res.data);
+      dispatch(login(username));
+      dispatch(loaded());
+    })
+    .catch((res) => {
+      dispatch(loginFail('Incorrect username or password'));
       dispatch(loaded());
     });
-  };
 };
-
-export const logoutThunk = () => {
-  return (dispatch) => {
-    dispatch(loading());
-    return axios
-      .get('/user/logout')
-      .then(() => {
-        dispatch(logout());
-        dispatch(loaded());
-      })
-      .catch((e) => {
-        dispatch(loaded());
-        console.log(e);
-      });
-  };
+export const whoami = () => (dispatch) => {
+  dispatch(loading());
+  return axios.get('/user/whoami').then(({ data }) => {
+    if (data.loggedIn) {
+      dispatch(login(data.username));
+    } else {
+      dispatch(logout());
+    }
+    dispatch(loaded());
+  });
 };
 
-export const getProductsThunk = () => {
-  return (dispatch) => {
-    dispatch(loading());
-    return axios
-      .get('/api/products')
-      .then(({ data }) => {
-        dispatch(getProducts(data));
-        dispatch(loaded());
-      })
-      .catch((e) => {
-        console.error(e);
-        dispatch(loaded());
-        return 'Error fetching products';
-      });
-  };
+export const logoutThunk = () => (dispatch) => {
+  dispatch(loading());
+  return axios
+    .get('/user/logout')
+    .then(() => {
+      dispatch(logout());
+      dispatch(loaded());
+    })
+    .catch((e) => {
+      dispatch(loaded());
+      console.log(e);
+    });
 };
 
-export const addProductThunk = (obj) => async (dispatch) => {
+export const getProductsThunk = () => (dispatch) => {
+  dispatch(loading());
   return axios
-    .post('/api/products', {
-      price: obj.price,
-      name: obj.name,
-      description: obj.description,
-      categoryId: obj.categoryId,
+    .get('/api/products')
+    .then(({ data }) => {
+      dispatch(getProducts(data));
+      dispatch(loaded());
     })
-    .then(() => axios.get('/api/products'))
+    .catch((e) => {
+      console.error(e);
+      dispatch(loaded());
+      return 'Error fetching products';
+    });
+};
+
+export const addProductThunk = (obj) => async (dispatch) => axios
+  .post('/api/products', {
+    price: obj.price,
+    name: obj.name,
+    description: obj.description,
+    categoryId: obj.categoryId,
+  })
```

Comment:
i think it would make more sense here if the post request returned what you're after when you call get on line 118.
seems like an unnecessary round trip to the server that could be done in one go.